include/sqlpp11/detail/parse_date_time.h

```diff
@@ -27,102 +27,237 @@
-        ::date::year{std::atoi(date_string + mr.position(1))} / // Year
-        std::atoi(date_string + mr.position(2)) /               // Month
-        std::atoi(date_string + mr.position(3))                 // Day of month
-      };
-      return true;
-    }
 
-    // Parse a date string formatted as YYYY-MM-DD HH:MM:SS.US TZ
-    // .US are optional fractional seconds, up to 6 digits in length
-    // TZ is an optional time zone offset formatted as +HH[:MM] or -HH[:MM]
-    //
-    inline bool parse_string_date_time(::sqlpp::chrono::microsecond_point& value, const char* date_time_string)
-    {
-      static const std::regex rx{
-        "(\\d{4})-(\\d{2})-(\\d{2}) "
-        "(\\d{2}):(\\d{2}):(\\d{2})(?:\\.(\\d{1,6}))?"
-        "(?:([+-])(\\d{2})(?::(\\d{2})(?::(\\d{2}))?)?)?"
-      };
-      std::cmatch mr;
-      if (std::regex_match(date_time_string, mr, rx) == false)
+      inline bool parse_character(const char*& input, char ch)
       {
-        return false;
+        if (*input != ch)
+        {
+          return false;
+        }
+        ++input;
+        return true;
       }
-      value =
-        ::sqlpp::chrono::day_point{
-          ::date::year{std::atoi(date_time_string + mr.position(1))} / // Year
-          std::atoi(date_time_string + mr.position(2)) /               // Month
-          std::atoi(date_time_string + mr.position(3))                 // Day of month
-        } +
-        std::chrono::hours{std::atoi(date_time_string + mr.position(4))} +     // Hour
-        std::chrono::minutes{std::atoi(date_time_string + mr.position(5))} +   // Minute
-        std::chrono::seconds{std::atoi(date_time_string + mr.position(6))} +   // Second
-        ::std::chrono::microseconds{                                           // Second fraction
-          mr[7].matched ? std::stoi((mr[7].str() + "000000").substr(0, 6)) : 0
-        };
-      if (mr[8].matched)
+
+      inline bool parse_yyyy_mm_dd(sqlpp::chrono::day_point& dp, const char*& input)
       {
-        const auto tz_sign = (date_time_string[mr.position(8)] == '+') ? 1 : -1;
-        const auto tz_offset =
-          std::chrono::hours{std::atoi(date_time_string + mr.position(9))} +
-          std::chrono::minutes{mr[10].matched ? std::atoi(date_time_string + mr.position(10)) : 0} +
-          std::chrono::seconds{mr[11].matched ? std::atoi(date_time_string + mr.position(11)) : 0};
-        value -= tz_sign * tz_offset;
+        auto new_input = input;
+        int year, month, day;
+        if ((parse_unsigned(year, new_input, 4) == false) || (parse_character(new_input, '-') == false) ||
+            (parse_unsigned(month, new_input, 2) == false) || (parse_character(new_input, '-') == false) ||
+            (parse_unsigned(day, new_input, 2) == false))
+        {
+          return false;
+        }
+        const auto ymd = ::date::year{year} / month / day;
+        // We could additionally check ymd.ok() but it is redundant as the above checks should suffice
```

Comment:
The checks above would allow for dates like `0000-00-00` or `2023-99-99`.

That 's probably OK, though, as the date would have come from the database this way.

It might therefore be more useful to document this in a more accessible place (the actual docs for instance).

Reply:
Initially there was a check `if (ymd.ok() == false)`, but the call to `ymd.ok()` turned out to be relatively slow and in debug builds the new manual parser code turned out to be about 25% slower than the old manual parser (which did not have the call to `ymd.ok()` anyway). In release builds the new parser was still much faster (about 10x - 3x faster than the old manual parser).

Given that the old manual parser did not have that check I though that we could drop that check in the new parser too.

However now that you say that we should document that I thought that it might be a problem worth documenting, I decided that it is probably worth adding the check anyway but speed it up somehow. So I wrote a custom implementation of the check for valid ymd combination and called the function `is_valid_ymd()`. It is significantly faster than `::date::year_month_day::ok`, so overall the new manual parser is still marginally (by about 1-2%) faster than the old manual parser in debug builds. In release builds it is 3-8 times faster than the old manual parser. I guess it makes sense to keep the check for valid ymd, given that it is fast enough. What do you think?

Reply:
> It is significantly faster than ::date::year_month_day::ok

Surprising. Maybe ask Howard Hinnant what he thinks about your implementation?

Reply:
This is the code of `date::year_month_day::ok()`
```
bool
year_month_day::ok() const NOEXCEPT
{
    if (!(y_.ok() && m_.ok()))
        return false;
    return date::day{1} <= d_ && d_ <= (y_ / m_ / last).day();
}
```
At first glance I can see at least 6 function calls (if we count `day{1}` as a constructor function call). Also `year_month_day_last::day()` internally does a bunch of function calls which further slows things down.

So I guess this is the reason why it is slow in debug, but it also seems to be slow in release builds with g++. Probably g++ doesn't optimize these function call very wells. In contrast the custom `is_valid_ymd()` only does 1 function call for YYYY-02-29 dates.

I guess I will need to write a simple test program that will compare the speed of `year_month_day::ok()` and the custom `is_valid_ymd()` and then if the speed difference remains I will contact Howard Hinnant. I am not sure that he will be inclined to optimize that part of the library though as I suspect that there are not that many applications that call `year_month_day::ok()` in a tight loop.

Also there is always the chance that I am missing something and the date library does the things correctly unlike my custom function.

Reply:
I'd still be interested in Howard's opinion regarding your version of checking validity, but that should not block this review...